Documentation/devicetree/bindings/media/i2c/sony,imx500.yaml

```diff
@@ -0,0 +1,161 @@
+  address of 0x36.  Image data is sent through MIPI CSI-2, which is configured
+  as either 2 or 4 data lanes.
+
+properties:
+  compatible:
+    const: sony,imx500
+
+  reg:
+    description: I2C device address
+    maxItems: 1
+
+  clocks:
+    maxItems: 1
+
+  clock-names:
+    description: |-
+      Input clock (12 to 27 MHz)
+    items:
+      - const: inck
+
+  interrupts:
+    maxItems: 1
+
+  vana-supply:
+    description: Supply voltage (analog) - 2.7 V
+
+  vdig-supply:
+    description: Supply voltage (digital) - 0.84 V
+
+  vif-supply:
+    description: Supply voltage (interface) - 1.8 V
+
+  reset-gpios:
+    description: |-
+      Sensor reset (XCLR) GPIO
+
+      Chip clear in lieu of built-in power on reset. To be set 'High' after
+      power supplies are brought up and INCK supplied.
+
+  port:
+    $ref: /schemas/graph.yaml#/$defs/port-base
+    additionalProperties: false
+    description: |
+      Video output port
+
+    properties:
+      endpoint:
+        $ref: /schemas/media/video-interfaces.yaml#
+        type: object
+        unevaluatedProperties: false
+        properties:
+          data-lanes:
+            items:
+              - const: 1
+              - const: 2
+          clock-noncontinuous: true
+          link-frequencies: true
+          clock-lanes:
+            items:
+              - const: 0
+        required:
+          - link-frequencies
+          - clock-lanes
+          - data-lanes
+
+required:
+  - compatible
+  - reg
+
+allOf:
+  - $ref: /schemas/spi/spi-peripheral-props.yaml#
+  - if:
+      required:
+        - spi-max-frequency
+    then:
+      properties:
+        # SPI node
+        clocks: false
+        clock-names: false
+        vana-supply: false
+        vdig-supply: false
+        vif-supply: false
+        reset-gpios: false
+        port: false
+    else:
+      required:
+        - clocks
+        - clock-names
+        - vana-supply
+        - vdig-supply
+        - vif-supply
+        - reset-gpios
+        - port
+
+unevaluatedProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+
+    i2c {
+      #address-cells = <1>;
+      #size-cells = <0>;
+
+      imx500: sensor@1a {
+        compatible = "sony,imx500";
+        reg = <0x1a>;
+
+        clocks = <&imx500_clk>;
+        clock-names = "inck";
+
+        vana-supply = <&imx500_vana>; /* 2.7  +/- 0.1  V */
+        vdig-supply = <&imx500_vdig>; /* 0.84 +/- 0.04 V */
+        vif-supply  = <&imx500_vif>;  /* 1.8  +/- 0.1  V */
+
+        reset-gpios = <&gpio_sensor 0 GPIO_ACTIVE_LOW>;
+
+        port {
+          imx500_0: endpoint {
+            clock-lanes = <0>;
+            remote-endpoint = <&csi1_ep>;
+            data-lanes = <1 2>;
+            clock-noncontinuous;
+            link-frequencies = /bits/ 64 <499500000>;
+          };
+        };
+      };
+    };
+
+    spi {
+      #address-cells = <1>;
+      #size-cells = <0>;
+
+      sensor@0 {
+        compatible = "sony,imx500";
```

Comment:
You've not pushed the driver yet, but doesn't this instantiate a second instance of the driver? How do the 2 get associated, or do they deliberately not?

I'm thinking of what happens if you have 2 attached to the system. How do you control which SPI and I2C device are linked, or does it not matter.

Reply:
I'm currently refactoring the driver but I will push it soon.

I'm doing something very similar to the s5c73m3 driver:
https://github.com/raspberrypi/linux/blob/5882bc5db17dd4a730faaedf547d073e487ec163/drivers/media/i2c/s5c73m3/s5c73m3-core.c#L1664
https://github.com/raspberrypi/linux/blob/5882bc5db17dd4a730faaedf547d073e487ec163/drivers/media/i2c/s5c73m3/s5c73m3-spi.c#L133

I'll have to examine the consequences of having 2 devices attached to the the system more closely.

Reply:
So AIUI that will gate registering the SPI driver part until the I2C bit has loaded, which is necessary as the i2c part allocates the common state structure.

However I don't see how that will work with two (it may not be applicable for that driver).
`s5c73m3_register_spi_driver` will get called when the first I2C device gets probed, but that will allow both SPI instances to probe, and the `container_of(spi->dev.driver, ...)` will point back to the one I2C instance.
When the second I2C device probes, `spi_register_driver` will get called for a second time, but I sort of hope it falls down the "Error: Driver '%s' is already registered, aborting" error path at https://elixir.bootlin.com/linux/latest/source/drivers/base/driver.c#L241

I don't know exactly what it would look like, but I would have expected a form not dissimilar to the `lens-focus` or `flash` DT links that you get from an image sensor to lens or flash devices. (Those get handled by the common v4l2-fwnode code, whereas this would be driver specific).